stdlib/dynarray.ml

```diff
@@ -102,17 +53,221 @@ and 'a slot =
    The following conditions define what we call a "valid" dynarray:
    - valid length: [length <= Array.length arr]
    - no missing element in the live space:
-     forall i, [0 <= i < length] implies [arr.(i) <> Empty]
+     forall i, [0 <= i < length] implies [arr.(i) != dummy]
    - no element in the empty space:
-     forall i, [length <= i < Array.length arr] implies [arr.(i) = Empty]
+     forall i, [length <= i < Array.length arr] implies [arr.(i) == dummy]
 
    Unfortunately, we cannot easily enforce validity as an invariant in
    presence of concurrent updates. We can thus observe dynarrays in
    "invalid states". Our implementation may raise exceptions or return
    incorrect results on observing invalid states, but of course it
    must preserve memory safety.
+
+   {3 Dummy values and flat float arrays}
+
+   OCaml performs a dynamic optimization of the representation of
+   float arrays, which is incompatible with our use of a dummy
+   value: if we initialize an array with user-provided elements,
+   it may get an optimized into a "flat float array", and
+   writing our non-float dummy values into it would crash.
+
+   To avoid interactions between unsafe dummy values and flat float
+   arrays, we ensure that the arrays that we use are never initialized
+   with floating point values. In that case we will always get
+   a non-flat array, and storing float values inside those is safe
+   (if less efficient). We call this the 'no-flat-float' invariant.
+
+   {3 Marshalling dummies}
+
+   There is a risk of interaction between dummies and
+   marshalling. If we use a global dynamically-allocated dummy
+   for the whole module, we are not robust to a user marshalling
+   a dynarray and unmarshalling it inside another program with
+   a different global dummy.
+
+   Instead, we create a dummy for each dynarray, and we store it in
```

Comment:
There could still be a global dummy that is used by all dynarrays but the unmarshalled ones, I think? That would save some allocations on each dynarray creation.

Reply:
I had the same idea while writing the description of the PR, and I pushed a commit to implement this -- but I forgot to update this bit of the documentation, thanks!